### ISSUE TITLE
Allow systemd socket passing for UNIX sockets (syslog source)

### ISSUE THREAD

**Issue by Korrigan** [vectordotdev/vector#19383]
### A note for the community

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Use Cases

In default configuration (On Debian at least), syslog daemons are expected to use the syslog.socket with the socket activation feature. This socket is a Unix datagram socket.
What I try to achieve is to use vector as my syslog daemon by using the socket passed by systemd.

### Attempted Solutions

Setting the mode to UDP allows socket passing but `listenfd.take_udp_socket` will fail because it checks the fd is indeed an UDP socket (which it's not because it's a UNIX datatagram socket, so we should use `listenfd.take_unix_datagram` instead in this case)

### Proposal

We should add a way to use FD passing with unix-type sockets (datagram or not)

I can see two possible approches here:
- Modify the TCP/UDP modes so they also accept UNIX sockets passed by systemd transparently (as users probably do not care if the underlying socket is an AF_INET or AF_UNIX type).
- Modify the "unix" mode behavior to accept sockets from systemd as well, although we also need it to accept datagram type sockets (maybe by introducing a "unix_datagram" mode ?)

I think the first option is the most reasonable and it avoids introducing a new "mode" and possibly more configuration options, If we agree on that, I'll try to come up with a PR.

Thanks

**Comment by dsmith3197** [comment]
Hi @Korrigan,

As you have encountered the `syslog` source does not currently have support for Unix Datagram sockets. However, you can use the `socket` source as an alternative ([ref](https://vector.dev/docs/reference/configuration/sources/socket/)).

You would set `decoding.codec = syslog` and `mode = unix_datagram`.

**Comment by dsmith3197** [comment]
This duplicates https://github.com/vectordotdev/vector/issues/8509.

Updated reference documentation here: https://github.com/vectordotdev/vector/pull/19399

**Comment by Korrigan** [comment]
Yes, Socket source sounds a great alternative as well but it doesn't seem the mode = "unix_datagram" supports systemd socket passing either, so maybe I could implement something like `mode = unix_datagram` + `path = systemd` (or `path = systemd#3`) to the socket source then ?

**Comment by dsmith3197** [comment]
The socket path is typically `/run/systemd/journal/syslog` on debian systems.

**Comment by Korrigan** [comment]
Yes absolutely! But I think my description of the use case might have been unclear, the concern I'm raising here is that the `path` option for `mode = unix_datagram` does not handle socket passing from systemd like the `address` option for `mode = tcp` for instance, it only allows to create our own sockets at the given `path`.
So I suggest to implement the same "systemd" and "systemd#N" magic values for `path` option on the socket source. It would work the same as using `mode = tcp` with `address = systemd`: We will use listenfd to take the underlying socket from systemd instead of creating our own.

I think I can tackle the implementation, but I'd rather make sure we agree on the feature before I do :)

**Comment by wjordan** [comment]
One possible workaround migh tbe to set your systemd unit's service config to [`StandardInput=socket`](https://www.freedesktop.org/software/systemd/man/latest/systemd.exec.html#StandardInput=), then use the `stdin` Vector source to receive the socket on the standard input fd.

**Comment by Jean-Daniel** [comment]
> One possible workaround might be to set your systemd unit's service config to [`StandardInput=socket`](https://www.freedesktop.org/software/systemd/man/latest/systemd.exec.html#StandardInput=), then use the `stdin` Vector source to receive the socket on the standard input fd.

That would be extremely inefficient, as it requires to launch a vector instance per connection to the socket.
